### DISCUSSION TITLE
Options and Swagger

### DISCUSSION THREAD

**Discussion by speterson-zoll** [louthy/language-ext#1132] [Q&A] [answered]
When I use `Option<string>` on a DTO, Swagger interprets the properties as arrays:

```
{
  "property1": [
    "string"
  ],
  "property2": [
    "string"
  ]
}
```

instead of

```
{
  "property1":  "string",
  "property2": "string"
}
```
If I use the second JSON, my API works just fine.

What I'm unclear about is where I need to deal with this.  Is there something I need to do in my LanguageExt code or is this strictly a Swagger issue?

**Comment by StefanBertels** [comment, marked as answer]
IMO:

Default JSON serialization of `Option` and other types is a bit different to what you might expect.  This is true for dictionary types, too.
I built my own custom serializer (for System.Text.Json) to make LanguageExt types more "compatible", i.e. serialize `Option<string>` in the same way as `string?`.

But this has some drawbacks and if you have a strict view on those things it is even better to serialize `Option` like an array because it is equivalent to an array that can have size 0 and size 1. So it depends who is consuming your API and how compatible you want/have to be.

**Reply by LaszloLueck** [reply]
Like in scala case classes and options, for me i wrote things like this:
```
Option<string> --> key : value
None --> key and value not in JSON

```
But if you do this, you must write your custom deserialization (for the way of json to your dto):

```
key, value not in json --> None
key, null in json --> None
key, value in json --> Option <string>
```

and so on.

Regards

**Reply by StefanBertels** [reply]
With Newtonsoft.Json I once wrote an ContractResolver (or something like that) that allowed to skip serialization of `None` values. This solution worked but I can't recommend this as it uses advanced details of Newtonsoft.Json. Anyway you should not use Newtonsoft.Json in new projects anymore (IMO), maybe get even rid if it (IMHO).

With System.Text.Json I had to use an ugly workaround to get the behaviour you specified because you cannot use a JsonConverter(Factory) to avoid serializing the property (meaning "not write the key"). Eventually this is a good thing as System.Text.Json serialization has a quite clean architecture.

What you can do: build a serializer that writes `null` for None. But you have to handle serializer settings, too (ignore value if default) if you want to avoid ugly JSON content containing "null". I ended up using a custom attribute to opt-in for this kind of serialization and another one to get rid of the item in JSON output when `None`: `[JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]`

Some issue with handling `Option<T>` like `T?` is that you have ambivalent values for use cases like `Option<Option<T>>`  (Some(None) looking like None).

**Reply by LaszloLueck** [reply]
Yes!
Thats what i mean with "and so on"

What if, when your Json contract say:

`key:value<number>`

and your json contains
`key:"abc"`


Your choice in your dto?
Exception (because of conversion error), handle as null and convert to none?
Second is not the best solution because you make a technical problem to a professional (in german i mean fachlich) result).

Indeed, what ist with `option<option<T>>`?

IMHO you can solve this "problem" to your own manner, there is no default way.

**Reply by StefanBertels** [reply]
I think having a string as a value for some numeric property has to throw an Exception and I guess it's this way for System.Text.Json with `int?`, too. Of course there are more difficult decisions like having string "123". There is a setting in System.Text.Json to deal with this and you should respect this. (Some other settings can be found here: https://learn.microsoft.com/de-de/dotnet/api/system.text.json.jsonserializeroptions). But this is IMO independent of LanguageExt types.

Regarding `Option<Option<T>>`: this is a valid type and for good reasons. LanguageExt enables us to express your intent using theses types without being limited by nesting or runtime problems. But probably having `Option<Option<T>>` is a semantically bad thing in an API because it is hard to understand and even in code I would try to avoid this, looks like a code smell. You probably want to use `Bind` somewhere or redesign your type signatures in your code and API.

Regardings custom JsonConverter(Factory): I personally don't want to handle special cases in general code and if you write a System.Text.Json.Serialization.JsonConverter(Factory) for some generic type you will struggle if you want to make this generic without array style JSON notation. And you should avoid to allow _reading_ different styles if there is a chance that this might have conflicts. If you allow "int[] style" and "int? style" at the same time for `Option<int>` then you might get conflicts when nesting types (array: inner or outer value).

Here's an Option<T> converter example: 

```
/// <summary>
/// serialize Option&lt;T&gt; like T? 
/// </summary>
/// <remarks>Add [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)] to the property to disable writing 'null' for None into json</remarks>
[AttributeUsage(AttributeTargets.Property | AttributeTargets.Field)]
public class OptionAsNullableJsonConverterAttribute : JsonConverterAttribute
{
    public override JsonConverter? CreateConverter(Type typeToConvert) => 
        Activator.CreateInstance(typeof(OptionJsonConverterFactory.OptionAsNullableJsonConverter<>).MakeGenericType(typeToConvert.IsGenericType(typeof(Option<>)) ? typeToConvert.GetGenericArguments().Head() : typeToConvert)) as JsonConverter;
}

internal class OptionJsonConverterFactory : JsonConverterFactory
{
    public override JsonConverter? CreateConverter(Type typeToConvert, JsonSerializerOptions options)
    {
        var innerType = typeToConvert.GetSingleGenericTypeArgument();
        return Activator.CreateInstance(typeof(OptionJsonConverter<>).MakeGenericType(innerType), options) as JsonConverter;
    }

    public override bool CanConvert(Type typeToConvert) => typeToConvert.IsGenericType(typeof(Option<>));

    private class OptionJsonConverter<T> : JsonConverter<Option<T>>
    {
        private readonly JsonConverter<T> _innerConverter;

        public OptionJsonConverter(JsonSerializerOptions options) => _innerConverter = (JsonConverter<T>) options.GetConverter(typeof(T));

        public override Option<T> Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
        {
            if (reader.TokenType == JsonTokenType.Null)
            {
                return None;
            }
            else
            {
                reader.Consume(JsonTokenType.StartArray);

                if (!reader.Read())
                {
                    throw new JsonException("could not read element in array or " + JsonTokenType.EndArray);
                }
                if (reader.TokenType != JsonTokenType.EndArray)
                {
                    // non-empty-array
                    var value = _innerConverter.Read(ref reader, typeof(T), options);
                    if (!reader.Read())
                    {
                        throw new JsonException("could not read element in array or " + JsonTokenType.EndArray);
                    }
                    if (reader.TokenType != JsonTokenType.EndArray)
                    {
                        throw new JsonException("expected token " + JsonTokenType.EndArray + " but found " + reader.TokenType);
                    }
                    return Option<T>.Some(value!);
                }
                else
                {
                    return Option<T>.None;
                }
            }
        }

        public override void Write(Utf8JsonWriter writer, Option<T> value, JsonSerializerOptions options)
        {
            writer.WriteStartArray();
            value.IfSome(some => _innerConverter.Write(writer, some, options));
            writer.WriteEndArray();
        }
    }

    internal class OptionAsNullableJsonConverter<T> : JsonConverter<Option<T>>
    {
        public override Option<T> Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options) =>
                reader.TokenType == JsonTokenType.Null ? None : Some((options.GetConverter(typeof(T?)) as JsonConverter<T> ?? throw new JsonException($"Could not get converter for {typeof(T)}")).Read(ref reader, typeof(T), options)!);

        public override void Write(Utf8JsonWriter writer, Option<T> value, JsonSerializerOptions options)
        {
            if (value.Case is T some)
            {
                ((JsonConverter<T>) options.GetConverter(typeof(T))).Write(writer, some, options);
            }
            else
            {
                writer.WriteNullValue();
            }
        }
    }
}
```

edit: some helper extensions I forgot

```
internal static Type GetSingleGenericTypeArgument(this Type genType) =>
    genType.GetGenericArguments().Match(
        () => throw new Exception("no generic type argument"),
        x => x, // fixed by edit, was `identity`
        (x, xs) => throw new Exception("more than one generic type argument")
    );

internal static bool IsGenericType(this Type t, Type genericTypeDefinition) => t.IsGenericType && t.GetGenericTypeDefinition() == genericTypeDefinition;
```

You can use it like this: `jsonSerializerOptions.Converters.Add(new OptionJsonConverterFactory());`

**Reply by speterson-zoll** [reply]
Thanks you very much for this example converter.   I am seeing two issues:

- On the line `reader.Consume(JsonTokenType.StartArray);`  The `Consume` method is not present on the `Utf8JsonReader` object.  What am I missing?
- Also, in the `GetSingleGenericTypeArgument` method in the extension, you pass in `identity`, but I see no other reference to that parameter.

**Reply by StefanBertels** [reply]
`identity` is in `Prelude`, so replace this with with `Prelude.identity` or add `using static LanguageExt.Prelude` at the top of your module. But then this is just equivalent to lambda `x => x` (identity function)... I will edit my answer to correct this inline.

`Consume` was a helper extension I forgot. This method just reads a token (consuming), verifies the type and (if different) throws an exception. Should be equivalent to `reader.Read()` with some `if(reader.TokenType != JsonTokenType.StartArray) throw new JsonException(...)`.

Sorry for incomplete code.

**Reply by speterson-zoll** [reply]
No problem; I appreciate what your quick responses!

I coded the `Consume` method like so:

```csharp
  internal static void Consume(this Utf8JsonReader reader, JsonTokenType jsonTokenType)
  {
      if (reader.TokenType != jsonTokenType)
      {
          throw new JsonException($"Could not read element in array or {jsonTokenType}");
      }
  }
```

One (hopefully) final issue:

When I wire up the `OptionJsonConverterFactory` as per your instructions, I no longer get a 500 exception, so it's already a large step ahead of the code I had.  Swagger still shows the properties as arrays:

```json
{
  "property1": [
    "string"
  ],
  "property2": [
    "string"
  ]
}
```

When I add the `OptionAsNullableJsonConverter` to one of the properties that can be null:

```csharp
       [OptionAsNullableJsonConverter]
        public Option<string> Property2 { get; set; }
```

and then run my code, I get back a 422 error:

```json
"errors": {
    "myDto": [
      "The myDto field is required."
    ],
    "$.proprty2": [
      "The JSON value could not be converted to LanguageExt.Option`1[System.String]. Path: $.property2 | LineNumber: 46 | BytePositionInLine: 23."
    ]
  }
```

Am I using the attribute correctly?

**Reply by StefanBertels** [reply]
Consume needs to consume a token, not just check. Use Read() before checking.

The attribute should work this way to get the non-array format. I use this in combination with [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)].

If this does not work, let me know. I hope I can back to this on Monday.

**Reply by speterson-zoll** [reply]
Thanks again.  I'm not that familiar with JS?ON readers, so I don't think I have the `Consume` method quite right.

Also, the addition of the `[JsonIgnore]` attribute did not alter the 422 error I received above.

**Reply by StefanBertels** [reply]
Sorry, there was a bug in this converter (for array style) regarding (my) Consume method. Here are some tests and an updated converted. (I don't know what is wrong in your setup, you should write a unit test or compare with my test methods.)

Tests:
```cs
public class SomeTypeWithOptions
{
    public Option<int> MaybeIntPropArray { get; set; }
    
    [OptionAsNullableJsonConverter]
    public Option<int> MaybeIntPropNullable { get; set; }

    [OptionAsNullableJsonConverter]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]
    public Option<int> MaybeIntPropNullableIgnore { get; set; }

}

[Fact]
void OptionTest1()
{
    var json = @"{ ""MaybeIntPropArray"": [4], ""MaybeIntPropNullable"": 5, ""MaybeIntPropNullableIgnore"": 6 }";

    var options = new JsonSerializerOptions(JsonSerializerDefaults.Web);
    options.Converters.Add(new OptionJsonConverterFactory());

    var result = JsonSerializer.Deserialize<SomeTypeWithOptions>(json, options);

    Assert.Equal(4, result!.MaybeIntPropArray);
    Assert.Equal(5, result!.MaybeIntPropNullable);
    Assert.Equal(6, result!.MaybeIntPropNullableIgnore);
}

[Fact]
void OptionTest2()
{
    var json = @"{}";

    var options = new JsonSerializerOptions(JsonSerializerDefaults.Web);
    options.Converters.Add(new OptionJsonConverterFactory());

    var result = JsonSerializer.Deserialize<SomeTypeWithOptions>(json, options);

    Assert.Equal(None, result!.MaybeIntPropArray);
    Assert.Equal(None, result!.MaybeIntPropNullable);
    Assert.Equal(None, result!.MaybeIntPropNullableIgnore);
}


[Fact]
void OptionTest3()
{
    var data = new SomeTypeWithOptions
    {
        MaybeIntPropArray =1,
        MaybeIntPropNullable = 2,
        MaybeIntPropNullableIgnore = 3
    };
    
    var options = new JsonSerializerOptions(JsonSerializerDefaults.General);
    options.Converters.Add(new OptionJsonConverterFactory());

    var json = JsonSerializer.Serialize(data, options);

    Assert.Equal(@"{""MaybeIntPropArray"":[1],""MaybeIntPropNullable"":2,""MaybeIntPropNullableIgnore"":3}", json);
}

[Fact]
void OptionTest4()
{
    var data = new SomeTypeWithOptions(); // all none
    var options = new JsonSerializerOptions(JsonSerializerDefaults.General);
    options.Converters.Add(new OptionJsonConverterFactory());

    var json = JsonSerializer.Serialize(data, options);

    Assert.Equal(@"{""MaybeIntPropArray"":[],""MaybeIntPropNullable"":null}", json);
}
```

Code:
```cs
/// <summary>
/// serialize Option&lt;T&gt; like T? 
/// </summary>
/// <remarks>Add [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)] to the property to disable writing 'null' for None into json</remarks>
[AttributeUsage(AttributeTargets.Property | AttributeTargets.Field)]
public class OptionAsNullableJsonConverterAttribute : JsonConverterAttribute
{
    public override JsonConverter? CreateConverter(Type typeToConvert) => 
        Activator.CreateInstance(typeof(OptionJsonConverterFactory.OptionAsNullableJsonConverter<>).MakeGenericType(typeToConvert.IsGenericType(typeof(Option<>)) ? typeToConvert.GetGenericArguments().Head() : typeToConvert)) as JsonConverter;
}

internal class OptionJsonConverterFactory : JsonConverterFactory
{
    public override JsonConverter? CreateConverter(Type typeToConvert, JsonSerializerOptions options)
    {
        // var innerType = typeToConvert.GetSingleGenericTypeArgument();
        var innerType = typeToConvert.GetGenericArguments().Match(
            () => throw new Exception("no generic type argument"),
            x => x,
            (x, xs) => throw new Exception("more than one generic type argument")
        );
        return Activator.CreateInstance(typeof(OptionJsonConverter<>).MakeGenericType(innerType), options) as JsonConverter;
    }

    public override bool CanConvert(Type typeToConvert) => typeToConvert.IsGenericType(typeof(Option<>));

    private class OptionJsonConverter<T> : JsonConverter<Option<T>>
    {
        private readonly JsonConverter<T> _innerConverter;

        public OptionJsonConverter(JsonSerializerOptions options) => _innerConverter = (JsonConverter<T>) options.GetConverter(typeof(T));

        public override Option<T> Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
        {
            if (reader.TokenType == JsonTokenType.Null)
            {
                return None;
            }
            else
            {
                if (reader.TokenType != JsonTokenType.StartArray)
                {
                    throw new JsonException($"expected {JsonTokenType.StartArray} but found {reader.TokenType}");
                }
                if (!reader.Read())
                {
                    throw new JsonException($"could not read element in array or {JsonTokenType.EndArray}");
                }
                if (reader.TokenType != JsonTokenType.EndArray)
                {
                    // non-empty-array
                    var value = _innerConverter.Read(ref reader, typeof(T), options);
                    if (!reader.Read())
                    {
                        throw new JsonException("could not read next token");
                    }
                    if (reader.TokenType != JsonTokenType.EndArray)
                    {
                        throw new JsonException($"expected token {JsonTokenType.EndArray} but found {reader.TokenType}");
                    }
                    return Option<T>.Some(value!);
                }
                else
                {
                    return Option<T>.None;
                }
            }
        }

        public override void Write(Utf8JsonWriter writer, Option<T> value, JsonSerializerOptions options)
        {
            writer.WriteStartArray();
            value.IfSome(some => _innerConverter.Write(writer, some, options));
            writer.WriteEndArray();
        }
    }

    internal class OptionAsNullableJsonConverter<T> : JsonConverter<Option<T>>
    {
        public override Option<T> Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options) =>
                reader.TokenType == JsonTokenType.Null ? None : Some((options.GetConverter(typeof(T?)) as JsonConverter<T> ?? throw new JsonException($"Could not get converter for {typeof(T)}")).Read(ref reader, typeof(T), options)!);

        public override void Write(Utf8JsonWriter writer, Option<T> value, JsonSerializerOptions options)
        {
            if (value.Case is T some)
            {
                ((JsonConverter<T>) options.GetConverter(typeof(T))).Write(writer, some, options);
            }
            else
            {
                writer.WriteNullValue();
            }
        }
    }
}
```

Helper extension method
```cs
internal static bool IsGenericType(this Type t, Type genericTypeDefinition) => t.IsGenericType && t.GetGenericTypeDefinition() == genericTypeDefinition;
```

**Reply by speterson-zoll** [reply]
Thank you!  Your updated code works well in terms of the unit tests passing.  However, something may be amiss in my Startup.cs in ASP.NET which is setup like so:

```csharp
            services.AddControllers().AddJsonOptions(x =>
            {
                x.JsonSerializerOptions.Converters.Add(new OptionJsonConverterFactory());
            });
```

I still get the same 422 error I received before whenever a property has the `[OptionAsNullableJsonConverter]` attribute on it.  Any thoughts about how this needs to be configured in Startup?  Debugging has not given me a definite path to follow.

**Reply by StefanBertels** [reply]
I tried this class in one of my Web apis (ASP.NET core) and it workes fine for me -- both as a parameter ([FromBody]) and as return type.

```
    public class EchoData
    {
        [OptionAsNullableJsonConverter]
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]
        public Option<string> MaybeString { get; set; }
    }
```

I suggest you check your serialization (what is sent to the server) and which JsonSerializerOptions you are using (your settings code above looks ok to me).
